pkg/descheduler/evictions/evictions.go

```diff
@@ -61,6 +63,7 @@ func NewPodEvictor(
 		client:             client,
 		policyGroupVersion: policyGroupVersion,
 		dryRun:             dryRun,
+		DegradationAllowed: degradationAllowed,
```

Comment:
`DegradationAllowed` field is not used anywhere inside PodEvictor methods. Is this PR still WIP?

Reply:
No, it's used by the eviction strategy, which is passed in the pod evictor. I opted to make this exportable so we could access the flag there, rather than having to modify every single eviction strategy callsite to add in another argument.

Reply:
I see: `podEvictor.DegradationAllowed`. It's not a good practice. 

> rather than having to modify every single eviction strategy callsite to add in another argument

However, it's the right thing to do from the maintainability perspective. If you need to access DegradationAllowed through `PodEvictor`, introducing new method is more practical than accessing `DegradationAllowed` directly. Also, `DegradationAllowed` is not related to evicting pod itself, rather to relaxing constraint of selecting pods to be evicted.

Reply:
I would argue that changing every function every time we add a new flag is worse from a maintainability point of view, but it's not my code base, so I'm happy to follow whatever convention people like. In terms of the degradation, you are correct, we are presently doing this through a global flag and applying it across all pods that satisfy the criteria. We could also consider pushing this down to the pod level through an additional annotation?

Reply:
> I would argue that changing every function every time we add a new flag

You are right. The `strategyFunction` data type is still evolving. I am fine with changing the signature [1] to:
```
type strategyFunction func(
	ctx context.Context,
	client clientset.Interface,
	strategy api.DeschedulerStrategy,
	nodes []*v1.Node,
	opts Options,
	podEvictor *evictions.PodEvictor,
)
``` 
where `Options` can be defined as:
```
type Options struct {
	EvictLocalStoragePods bool
	DegradationAllowed bool
}
```
as long as `Options` contains fields generic for any strategy.

[1] https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/descheduler/descheduler.go#L63

> We could also consider pushing this down to the pod level through an additional annotation?

If I understand the intention correctly, `DegradationAllowed` is a strategy level configuration.  As such, the option needs to be passed into the strategy before it's ran.

Reply:
The problem with this is that the descheduler package would define the Options struct, while the strategies package would need to import it for the function definitions, and the descheduler package already imports the strategies package, so we would end up with a circular import. I've worked around this by placing the Options struct into the strategies package. Should we want to use this in e.g. the pod package for option evaluation within ListEvictablePodsOnNode(), however, we will again run into a circular import.

I will open a separate PR for the options refactoring, and once we're happy with that I'll then update this PR.

Reply:
Please refer to PR #305

Reply:
@pmundt if this is only used by the `nodeAffinity` (and, in your new strategy `nodeSelector`), does it have to be a global setting? It could just be a [`StrategyParam`] field for those right?

Reply:
As noted in the other PR (sorry for the run-around, I mention it here for posterity), we would also plan to evaluate it within `IsEvictable` in cases where we know the pod can be explicitly degraded. I'm therefore not sure that the `StrategyParam` would be sufficient, unless we're also able to access this from the pod package somehow. In terms of the global setting, I don't mind if we get rid of this and use a Pod annotation or similar, the important thing is simply that we have a mechanism to degrade node-local Pods - I'll defer to you on whichever option you find more palatable! If we leave it as an annotation, we could presumably also leave it as a `StrategyParam`, as the annotation could be tested independently within `IsEvictable`.

Reply:
> In terms of the global setting, I don't mind if we get rid of this and use a Pod annotation or similar, the important thing is simply that we have a mechanism to degrade specific node-local Pods

Is this something that the existing [`descheduler.alpha.kubernetes.io/evict`](https://github.com/kubernetes-sigs/descheduler/blob/bd412bf87f72132efad2a07833666463cb909969/pkg/descheduler/pod/pods.go#L32) annotation could solve? It is already checked in `IsEvictable` and effectively bypasses that check for specific pods.

> If we leave it as an annotation, we could presumably also leave it as a StrategyParam, as the annotation could be tested independently within IsEvictable.

With the above annotation^ I think this is what you want. In [the other PR thread](https://github.com/kubernetes-sigs/descheduler/pull/305#issuecomment-637463597) you mentioned wanting to specifically bypass DaemonSets too, which that annotation supports.

Reply:
> Is this something that the existing [`descheduler.alpha.kubernetes.io/evict`](https://github.com/kubernetes-sigs/descheduler/blob/bd412bf87f72132efad2a07833666463cb909969/pkg/descheduler/pod/pods.go#L32) annotation could solve? It is already checked in `IsEvictable` and effectively bypasses that check for specific pods.

You're right, for some reason when I first took a look at the evict annotation I missed this. I think this would do the job, yes.
 
> With the above annotation^ I think this is what you want. In [the other PR thread](https://github.com/kubernetes-sigs/descheduler/pull/305#issuecomment-637463597) you mentioned wanting to specifically bypass DaemonSets too, which that annotation supports.
> 
> Sorry if it seems like I'm being difficult, I just don't yet see the need to pass this information to every strategy.

No problem, it's not always obvious what the preferred direction is when twiddling in someone else's code base, and as it turns out, I misread what the evict annotation actually does, so I'm happy for another set of eyes while I come to grips with things.

If you're happy with the `StrategyParam` direction, I'll give this a go with the evict annotation and see how it goes.

Reply:
Yes let's do that -- I agree that there are cases when these affinity strategies may want to evict even if the pod can't fit anywhere else, so adding another Param for these strategies makes sense to me.

I'm also assuming that if you enable that Param the user must know what they are doing, because as @ingvagabund [pointed out](https://github.com/kubernetes-sigs/descheduler/pull/301#discussion_r433909302) if you're not properly handling such an evict-at-all-costs case you will end up with stuck Pending pods. I don't think that's the descheduler's concern though (as long as the possibility is clearly documented)